### PR TITLE
Sony Smartwatch 3 (tetra) support

### PR DESCRIPTION
**Pull request by FlorentRevest**
This Pull Requet includes an initial merging effort of the AOSP's bcm kernel for the tetra platform. This is still a work in progress and comments are welcome and needed.

I tried to stick as close as possible to the current sonyxperiadev's code. For example I tried to merge the bcm's drivers/gpu/ion/bcm into drivers/staging/android/ion but it probably added problems so we should undo those changes.

The current code compiles but it doesn't boot. I still haven't succeeded at getting apanic logs.

### CODE REVIEW

**Comment by kholk** [comment]
I suppose we should try to get some UART output from that thing.
That's some essential thing when you actually try to bringup an ENTIRE PLATFORM with even a NEW CPU on a certain kernel... :\

Your bringup seems to be pretty clean, but I've seen that there are some commits that are being reverted later. I would remove both the commit and its revert to keep the commitlog clean: remember that this kernel supports lots of devices!

File: .gitignore [hunk 1]
```diff
@@ -32,6 +32,7 @@ modules.builtin
 *.lzo
 *.patch
 *.gcno
+out/
```
Comment (kholk) [inline .gitignore]: While I'd like to get an out folder, this should not be in gitignore, as building with outfolder is not a default kernel compile "option".

File: .gitignore [hunk 2]
```diff
@@ -74,6 +75,19 @@ series
 cscope.*
 ncscope.*
 
+# Gator daemon Files
```
Comment (kholk) [inline .gitignore]: Those ones I would put them at the bottom of the gitignore file, along with the other Broadcom ignores, to still not diverge too much from QC.

File: arch/arm/Kconfig [hunk 3]
```diff
@@ -1743,6 +1778,17 @@ config ARCH_SELECT_MEMORY_MODEL
 config HAVE_ARCH_PFN_VALID
 	def_bool ARCH_HAS_HOLES_MEMORYMODEL || !SPARSEMEM
 
+config ARCH_ENABLE_MEMORY_HOTPLUG
```
Comment (kholk) [inline arch/arm/Kconfig]: Can potentially break QC support. If it is a Broadcom-specific config option (along with the other two), then set def_bool y only if BCM chips.

File: arch/arm/Makefile [hunk 4]
```diff
@@ -124,8 +124,24 @@ AFLAGS_ISA	:=$(CFLAGS_ISA)
 endif
 
 # Need -Uarm for gcc < 3.x
-KBUILD_CFLAGS	+=$(CFLAGS_ABI) $(CFLAGS_ISA) $(arch-y) $(tune-y) $(call cc-option,-mshort-load-bytes,$(call cc-option,-malignment-traps,)) -msoft-float -Uarm
-KBUILD_AFLAGS	+=$(CFLAGS_ABI) $(AFLAGS_ISA) $(arch-y) $(tune-y) -include asm/unified.h -msoft-float
+KBUILD_CFLAGS	+=$(CFLAGS_ABI) $(CFLAGS_ISA) $(arch-y) $(tune-y) $(call cc-option,-mshort-load-bytes,$(call cc-option,-malignment-traps,)) -Uarm
+KBUILD_AFLAGS	+=$(CFLAGS_ABI) $(AFLAGS_ISA) $(arch-y) $(tune-y) -include asm/unified.h
+
+ifeq ($(CONFIG_VFP),y)
```
Comment (kholk) [inline arch/arm/Makefile]: We have VFPv3 on QC, yet we are compiling with softfloat. Can this break anything OR cause performance issues? hmmmm :)

File: arch/arm/Makefile [hunk 5]
```diff
@@ -288,11 +309,25 @@ boot := arch/arm/boot
 
 archprepare:
 	$(Q)$(MAKE) $(build)=arch/arm/tools include/generated/mach-types.h
+ifeq ($(CONFIG_ARCH_RHEA_BX),y)
```
Comment (kholk) [inline arch/arm/Makefile]: Don't we have any clean ways for this? We surely do!!!!

File: arch/arm/boot/Makefile [hunk 6]
```diff
@@ -32,7 +35,27 @@ INITRD_PHYS := $(initrd_phys-y)
 
 export ZRELADDR INITRD_PHYS PARAMS_PHYS
 
-targets := Image zImage xipImage bootpImage uImage
+targets := Image zImage xipImage bootpImage uImage dt-blob
+
+$(obj)/dt-blob: FORCE
```
Comment (kholk) [inline arch/arm/boot/Makefile]: Remove that dt-blob crap!!!!! Do it cleanly!!!

Bloating an already bloated QC kernel with bloaty bloat is... bloaty? :)

File: arch/arm/boot/bootp/Makefile [hunk 7]
```diff
@@ -5,6 +5,11 @@
 # architecture-specific flags and dependencies.
 #
 
+# GCOV patch
```
Comment (kholk) [inline arch/arm/boot/bootp/Makefile]: Qualcomm supports that. Why do we have that global hack?

File: arch/arm/boot/compressed/Makefile [hunk 8]
```diff
@@ -29,6 +29,10 @@ OBJS	+= debug.o
 endif
 FONTC	= $(srctree)/drivers/video/console/font_acorn_8x8.c
 
+# GCOV patch
```
Comment (kholk) [inline arch/arm/boot/compressed/Makefile]: Qualcomm supports that. Why do we have that global hack?

File: arch/arm/boot/compressed/Makefile [hunk 9]
```diff
@@ -148,7 +152,7 @@ LDFLAGS_vmlinux += -X
 LDFLAGS_vmlinux += -T
 
 # For __aeabi_uidivmod
-lib1funcs = $(obj)/lib1funcs.o
+lib1funcs = $(obj)/lib1funcs.o $(obj)/divdi3.o
```
Comment (kholk) [inline arch/arm/boot/compressed/Makefile]: Looks like we don't really need it on any platform we currently have. Will it break? whoooooooo knows :)

File: arch/arm/boot/compressed/head.S [hunk 10]
```diff
@@ -687,9 +687,70 @@ __armv4_mmu_cache_on:
 		mcr	p15, 0, r0, c8, c7, 0	@ flush I,D TLBs
 #endif
 		mov	pc, r12
+#ifdef CONFIG_KONA_DCACHE_INVALIDATION
```
Comment (kholk) [inline arch/arm/boot/compressed/head.S:690]: Do we have Kona? Isn't Smartwatch 3 Java arch?

File: arch/arm/boot/dts/Hawaii.dts [hunk 11]
```diff
@@ -0,0 +1,210 @@
+/dts-v1/;
```
Comment (kholk) [inline arch/arm/boot/dts/Hawaii.dts]: Can we stay clean, again, and remove anything that is not MACH_JAVA and Sony stuff?
Maybe it'd be okay if we don't support other old Broadcom CPUs :)

File: arch/arm/boot/dts/Makefile [hunk 12]
```diff
@@ -207,6 +207,41 @@ dtb-$(CONFIG_ARCH_VT8500) += vt8500-bv07.dtb \
 	wm8650-mid.dtb \
 	wm8850-w70v2.dtb
 dtb-$(CONFIG_ARCH_ZYNQ) += zynq-zc702.dtb
+dtb-$(CONFIG_ARCH_HAWAII) += hawaii_garnet_edn020.dtb \
```
Comment (kholk) [inline arch/arm/boot/dts/Makefile]: Same here, maybe remove even ARCH_JAVA DTs, keep only SONY stuff.
...Or maybe not, your choice here

File: arch/arm/boot/dts/Hawaii.dts (hunk 11 above)
Comment (FlorentRevest) [inline arch/arm/boot/dts/Hawaii.dts]: Both CONFIG_ARCH_KONA https://android.googlesource.com/kernel/bcm/+/android-bcm-tetra-3.10-lollipop-mr1-wear-release/arch/arm/configs/tetra_defconfig#36 and CONFIG_ARCH_JAVA https://android.googlesource.com/kernel/bcm/+/android-bcm-tetra-3.10-lollipop-mr1-wear-release/arch/arm/configs/tetra_defconfig#45 are set in tetra_defconfig so I believe those two archs are needed. Anyway, I'm carefully reading all of your comments and this is really great, thanks for your time and help.

**Comment by FlorentRevest** [comment]
Sorry for the late answer, I had a busy month. However I come with great news, indeed:
- All the problems you underlined have been fully fixed.
- The codebase is also much cleaner than before
- It now boots really well. (last time it wasn't working)
- Bcm's Ion has been removed at the request of Alin Jerpelea so for now there is no graphics but adding ion makes GUI work.

Only the following files haven't been correctly merged, yet:
- drivers/usb/gadget/Kconfig
- drivers/usb/gadget/android.c
- drivers/usb/gadget/composite.c
- drivers/usb/gadget/f_acm.c
- drivers/usb/gadget/f_hid.c
- drivers/usb/gadget/f_mass_storage.c
- drivers/usb/gadget/f_ncm.c
- drivers/usb/gadget/f_rndis.c
- drivers/usb/gadget/f_serial.c
- drivers/usb/gadget/rndis.c
- drivers/usb/gadget/rndis.h
- drivers/usb/gadget/storage_common.c
- drivers/usb/gadget/u_ether.c
- drivers/usb/gadget/u_ether.h
- drivers/usb/gadget/u_serial.c
- drivers/usb/gadget/u_serial.h
- drivers/usb/gadget/udc-core.c

They are tightly linked and boot logs don't give any clue about what's going wrong so it's hard to find the root of the issue. While I'm working on them I am open to any comment you have!

**Comment by FlorentRevest** [comment]
USB support is now fixed, so this should be ready for a merge. I'm waiting for your comments!

**Comment by kholk** [comment]
is the eMMC working?

**Comment by FlorentRevest** [comment]
Yes it does. adb shell is available to try to modify things on disk.

**Pull request merged**